server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java

```diff
@@ -432,4 +444,39 @@ public String toString() {
             }
         };
     }
+
+    /**
+     * Determines whether vector fields should be excluded from the source based on the {@link FetchSourceContext}.
+     * Returns {@code true} if vector fields are explicitly marked to be excluded and {@code false} otherwise.
+     */
+    private static boolean shouldExcludeVectorsFromSource(SearchContext context) {
+        if (context.fetchSourceContext() == null) {
+            return false;
+        }
+        return context.fetchSourceContext().includeVectors() != null && context.fetchSourceContext().includeVectors() == false;
+    }
+
+    /**
+     * Returns a {@link SourceFilter} that excludes vector fields not associated with semantic text fields,
+     * unless vectors are explicitly requested to be included in the source.
+     * Returns {@code null} when vectors should not be filtered out.
+     */
+    private static SourceFilter maybeExcludeNonSemanticTextVectors(SearchContext context) {
+        if (shouldExcludeVectorsFromSource(context) == false) {
+            return null;
+        }
+        var lookup = context.getSearchExecutionContext().getMappingLookup();
+        List<String> inferencePatterns = lookup.inferenceFields().isEmpty() ? null : lookup.inferenceFields().keySet().stream().toList();
+        var excludes = lookup.getFullNameToFieldType()
+            .values()
+            .stream()
+            .filter(
+                f -> f instanceof DenseVectorFieldMapper.DenseVectorFieldType || f instanceof SparseVectorFieldMapper.SparseVectorFieldType
```

Comment:
it seems to me that this should be a general "mapping.type" exclusion. I realize this complicates things with plugin mapper fields (rank_vector...), but rank_vectors should also be included here...